### PR TITLE
[Merged by Bors] - chore: remove `autoImplicit` in `Tactic.Ring.Basic`

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Deprecated/HashMap.lean [hunk 1]
```diff
@@ -17,13 +17,13 @@ import Batteries.Data.RBMap.Basic
...
 
-set_option autoImplicit true
+universe u v
 
 namespace Batteries.HashMap
 
 -- not an exact match, the Lean3 version was dependently-typed
 
-variable [BEq α] [Hashable α]
+variable {α : Type u} {β : Type v} [BEq α] [Hashable α]
```
Comment (grunweg) [inline Mathlib/Deprecated/HashMap.lean:26]: Can you use Type* here, and/or move \alpha outside of the namespace?

**Review by grunweg — COMMENTED** [review]
Thanks for looking into this! Quite a few of these changes seem positive to me. For example, I like how the Data/Vector changes turned out, they feel much cleaner than #14885.
- I'm on the fence about some files (say: Cache, the archive or deprecated files don't feel as important - one could also leave them alone). That said, the archive and cache changes are quite neutral/a net negative in lines of code, so I can certainly live with them.
- Control/Random in fact uses autoImplicit pervasively: I am not sure if changing this file is a net positive. (The most confusing feature of autoImplicit is if *some* declarations use it, but not all. When you change 5 out of 100 declarations, the consistency feels like a clear improvement. As the entire file seems to use it, I'm less convinced. Adding a comment that this is on purpose would also work.)
- Can you split Equiv/Basic into a separate PR, as that file is simply large enough?
- I am not convinced about the changes to `ToExpr`.

File: Mathlib/Control/Random.lean [hunk 2]
```diff
@@ -130,17 +134,17 @@ instance {n : Nat} : BoundedRandom m (Fin n) where
...
-    let ⟨x⟩ ← ULiftable.up.{v} (BoundedRandom.randomR lo.down hi.down h)
+    let ⟨x⟩ ← ULiftable.up (BoundedRandom.randomR lo.down hi.down h)
     pure ⟨ULift.up x.val, x.prop⟩
 
 end Random
 
 namespace IO
 
-variable {m : Type* → Type*} {m₀ : Type → Type}
+variable {m : Type u → Type*} {m₀ : Type → Type}
```
[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Control/Random.lean:147]: `RandT m α` depends on `StateT m α`, the latter of which requires that the first universe in `{m : Type u → Type*}` match the universe of `α`.

File: Mathlib/Tactic/Ring/Basic.lean [hunk 3]
```diff
@@ -164,28 +166,34 @@ mutual -- partial only to speed up compilation
...
+    ExProd sα a → ExProd sα b → Ordering
   | .const i _, .const j _ => compare i j
   | .mul a₁ a₂ a₃, .mul b₁ b₂ b₃ => (a₁.cmp b₁).then (a₂.cmp b₂) |>.then (a₃.cmp b₃)
   | .const _ _, .mul .. => .lt
   | .mul .., .const _ _ => .gt
 
 @[inherit_doc ExBase.cmp]
-partial def ExSum.cmp : ExSum sα a → ExSum sα b → Ordering
+partial def ExSum.cmp
+    {u : Lean.Level} {α : Q(Type u)} {sα : Q(CommSemiring «$α»)} {a b : Q(«$α»)} :
```
[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Tactic/Ring/Basic.lean:188]: It doesn't work. When I make `variable {u : Lean.Level} {α : Q(Type u)}` I get the error
```
application type mismatch
  ExProd.eq ?m.8597 b₂
argument
  b₂
has type
  @ExProd Lean.Level.zero (Expr.const `Nat []) sℕ e✝ : Type
but is expected to have type
  @ExProd u α sℕ e✝ : Type
```

File: Mathlib/Tactic/ToExpr.lean [hunk 4]
```diff
@@ -21,13 +21,28 @@ namespace Lean
 
 attribute [-instance] Lean.instToExprOption
 
-set_option autoImplicit true in
-deriving instance ToExpr for Option
+universe u in
```
[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Tactic/ToExpr.lean:24]: Actually it''s quite logical. Foe the first three instances I copied code over from core and made it universe-polymorphic. The fourth explicitly written out instance follows the style of the first three.

File: Mathlib/Logic/UnivLE.lean [hunk 5]
```diff
@@ -53,7 +53,7 @@ theorem UnivLE.trans [UnivLE.{u, v}] [UnivLE.{v, w}] : UnivLE.{u, w} :=
   fun α ↦ Small.trans_univLE α
 
 /- This is the crucial instance that subsumes `univLE_max`. -/
-instance univLE_of_max [UnivLE.{max u v, v}] : UnivLE.{u, v} := @UnivLE.trans univLE_max.{v,u} ‹_›
+instance univLE_of_max [UnivLE.{max u v, v}] : UnivLE.{u, v} := @UnivLE.trans univLE_max.{u,v} ‹_›
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Logic/UnivLE.lean:56]: Why do you need this change?

[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Logic/UnivLE.lean:56]: It doesn't work the other way.

[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Logic/UnivLE.lean:56]: Please pull this change out into a separate PR. It seems unrelated?

But yes, it seems to be correct, and the universe ordering on `univLE_max` was "backwards" previously.

[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Logic/UnivLE.lean:56]: @semorrison #16690

**Comment by grunweg** [comment]
I took a look again: the changes to `Ring/Basic.lean` are not pretty, but I can certainly live with them (in exchange for not switching style half-way throughout the file).

I am not convinced by the readability regression in ToExpr. I tend to leave this alone, and perhaps add a comment instead.
Similarly, for Control/Random, my personal preferences is to have a comment instead.

I can mention this on zulip again (in a week or two; I have very little Lean time right now). I would maintainer merge the changes to Ring/Basic right away.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**